Automattic/woocommerce-payments
# Add/optout survey UI

Fixes #2311 

#### Changes proposed in this Pull Request

Create the UI elements for the following:

- Modal for confirming + disabling UPE.
- Modal for submitting feedback for UPE. This is sent to the Jetpack marketing surveys endpoint on Missioncontrol.

### Screencast
![Screen Capture on 2021-07-22 at 18-04-42](https://user-images.githubusercontent.com/6723003/126674648-ed31b46f-7347-4dc8-a960-0309c97c46fd.gif)

### Description

Before disabling UPE, we want to display a confirmation modal; it goes over which enabled payment methods would be rendered dysfunctional if UPE was disabled. Clicking the Disable button should disable UPE.

That should render a toast notification and immediately load the Survey Modal which prompts users to send feedback. Their selection and comments should be sent to the site's /upe-survey/ endpoint which goes to MissionControl on WPCOM through the /marketing/surveys endpoint.

### Testing

There are 2 paths to take:

- **Go directly to Provide Feedback.**

Make sure you look at the prerequisites for the POC in #2308, notably:

>Change your host file to point public-api.wordpress.com to your sandbox's IP address
>Connect to your sandbox via SSH

1. Make sure UPE is enabled.
2. Clicking on the hamburger menu within the Payment Methods section of the Settings page.
3. Select Provide Feedback.
4. Select one of the options and add a comment.

- Go to Disable, and after disabling UPE, be prompted with the Provide Feedback modal.

After the survey submission, go here while ssh'ed into your sandbox:

https://dev-mc.a8c.com/marketing/survey/?survey=wcpay-upe-disable-early-access&start=&end=&survey_filter_function=

You should see your answer tallied there and click on 'see answer details' for the comment you wrote:

### Comments

- I started refactoring the SurveyModal component to be more re-usable; it's not _quite_ there as it still has a hard-coded API fetch path(/upe_survey) in `WcPaySurveyContextProvider`. When more tests are added we could add another prop to `WcPaySurveyContextProvider` and pass the path in when using it. Let me know of any other opportunities for refactoring so it can be done.

- In an attempt to make the Survey Modal closer to re-usable, the survey questions are loaded from a static JS file named `questions.js`. It makes the mapping of data to react elements a bit tricky but you could add in additional surveys and re-use the component later with some slight adjustments.


-------------------
